pkg/expressions/function/functions.go

```diff
@@ -600,6 +805,77 @@ func getChart(
 	}
 }
 
+// getChartFromDiscoveredArtifacts returns a function that finds the latest Helm chart based on repository URL.
+func getChartFromDiscoveredArtifacts(
+	ctx context.Context,
+	wh *kargoapi.Warehouse,
+	artifacts *kargoapi.DiscoveredArtifacts,
+) exprFn {
+	return func(a ...any) (any, error) {
+		if len(a) != 1 {
+			return nil, fmt.Errorf("expected 1 argument, got %d", len(a))
+		}
+
+		repoURL, ok := a[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("first argument must be string, got %T", a[0])
+		}
+
+		repoURL = urls.NormalizeChart(repoURL)
+
+		logger := logging.LoggerFromContext(ctx).WithValues(
+			"repoURL", repoURL,
+			"warehouse", wh.Name,
+		)
+
+		var latestChart *kargoapi.Chart
+		for _, s := range wh.Spec.Subscriptions {
+			if s.Chart != nil && urls.NormalizeChart(s.Chart.RepoURL) == repoURL && len(artifacts.Charts) != 0 {
+				logger.Debug("number of discovered chart artifacts",
+					"count", len(artifacts.Charts),
+				)
+				for _, ca := range artifacts.Charts {
+					caRepoURL := urls.NormalizeChart(ca.RepoURL)
+					if caRepoURL != repoURL {
+						continue
+					}
+					for _, v := range ca.Versions {
+						sv, err := semver.NewVersion(v)
+						if err != nil {
+							logger.Error(err, "ignoring invalid semver version", "version", v)
+							continue
+						}
+						if latestChart == nil {
+							latestChart = &kargoapi.Chart{
+								RepoURL: repoURL,
+								Name:    caRepoURL,
+								Version: sv.String(),
+							}
+							continue
+						}
+						latestSemver, err := semver.NewVersion(latestChart.Version)
+						if err != nil {
+							logger.Error(err, "ignoring invalid semver version", "version", latestChart.Version)
+							continue
+						}
+						if sv.GreaterThan(latestSemver) {
+							latestChart = &kargoapi.Chart{
+								RepoURL: caRepoURL,
+								Name:    ca.Name,
+								Version: sv.String(),
+							}
+						}
+					}
```

Comment:
Similar question to [here](https://github.com/akuity/kargo/pull/5047/files#r2377074103) with an additional question about why semver...

The selection criteria may not have been semver based.

So as in the other case, we should trust that the discovery process ordered the artifacts correctly.